### ISSUE TITLE
Add  infix "!in" as synonym for ∉ ?

### ISSUE THREAD

**Issue by samoconnor** [JuliaLang/julia#15353]
I miss not having [infix `ni` as the opposite of `in` like Tcl has](https://www.tcl.tk/man/tcl8.6/TclCmd/expr.htm#M15)

`∉` is nice, but I don't know how to type that.

I end up with a lot of `if !(x in v) ...` which I'd rather replace with `if x ni v ...`
Does this seem crazy to people without a Tcl disorder?

_Update: waiting for general infix `!` per https://github.com/JuliaLang/julia/issues/15353#issuecomment-192480765 ._

**Comment by nalimilan** [comment]
?

**Comment by samoconnor** [comment]
Sorry @nalimilan I think I pressed enter too soon an submitted a half-typed issue.

**Comment by Dawny33** [comment]
`∉` can be typed by `\notin` (both in Atom with the latex plugin and the REPL)

**Comment by samoconnor** [comment]
HI @Dawny33, I'm not sure what Atom is. I mostly only ever type code into Vim. I'm sure there is a way to type `∉` in Vim, but I can only fit a finite number of Vim shortcuts in my brain.

I think it's reasonable to expect that non-ascii operators have an ascii-typeable equivalent.

**Comment by Dawny33** [comment]
@samoconnor Yeah, the request is quite reasonable :+1: 

Just trying to point out the latex shortcut for quick reference!

**Comment by nalimilan** [comment]
> Does this seem crazy to people without a Tcl disorder?

Yes, I must confess I'm in that camp. :-)  Using `if`/`fi` and things like that always sounded weird to me. We may as well use `out` (just joking).

**Comment by samoconnor** [comment]
`ni`  is nice because it is `reverse("in")` and also "**n**ot **i**n" and also because [Ni!](https://youtu.be/oas2l8MnyC0) 

But if `ni` is too cute 🐶, then `!in` or `not in` or `notin` would be fine too...

**Comment by tkelman** [comment]
If we really need an ascii infix not-in operator (is `!(a in b)` not sufficient for most purposes?) then I'd rather have it spelled `notin` than `ni`.

**Comment by Dawny33** [comment]
@tkelman +1

`notin` plays well with the latex shortcut `\notin`, so it avoids unnecessary confusion.  And yeah, we need a `not in` operator, as it is a bit of a frequent usage.

**Comment by KristofferC** [comment]
Ni! Ni! Ni!

**Comment by samoconnor** [comment]
@KristofferC Icky Icky Icky Ptang

**Comment by nalimilan** [comment]
Indeed, the movie reference is the only possible justification of Tcl's choice.

**Comment by samoconnor** [comment]
Adding `notin(x, itr) = !in(x,itr)` to reduce.jl is easy enough.
What has to be changed to make it work as an infix operator?

**Comment by eschnett** [comment]
Clearly, `ni` should be equivalent to `\ni, which is `∋` -- the left-right swapped version of `in`.

**Comment by vchuravy** [comment]
@samoconnor You might want to look at https://github.com/JuliaLang/julia-vim AFAIK it can replace \notin with the appropriate unicode symbol

**Comment by StefanKarpinski** [comment]
Perhaps parsing `!in` as infix would be better, as in `a !in b`.

**Comment by kmsquire** [comment]
+1 for `!in` as an infix operator.

**Comment by JeffBezanson** [comment]
One reason to prefer `notin` is that it's compatible with allowing all functions to use infix syntax, while `!in` would continue to be a special case even if we made that change.

**Comment by StefanKarpinski** [comment]
If `a foo b` desugars to `foo(a,b)` then `a !foo b` could generically desugar to `!foo(a,b)`. On the other hand, I find this particular syntax issue to be mild enough that it can wait.

**Comment by nalimilan** [comment]
Could be useful, but we'll have to wait until more infix operators are supported to really give `!in` that general meaning. Until then, it will only be a weird exception.

**Comment by samoconnor** [comment]
If it is agreed that a general infix `!` and `!in` can wait, then `notin` seems to be broadly agreeable.
Would a PR for `notin` be welcome?

**Comment by StefanKarpinski** [comment]
I'm against it because then we'd have to deprecate `notin` when `!in` happens. In the meantime, writing `!(a in b)` is not so awful.

**Comment by samoconnor** [comment]
Fair enough.

**Comment by KristofferC** [comment]
Now have `!in`

**Comment by DNF2** [comment]
Does it also solve https://github.com/JuliaLang/julia/issues/25512, or does it only work for `in`?

**Comment by nalimilan** [comment]
But `x !in y` doesn't work, right? Yet that's what this issue is about.

**Comment by KristofferC** [comment]
Yeah, sorry, I missed the actual point here.

**Comment by Moelf** [comment]
I think people are upset that you have to write `!in(2, [2,3])` but can't write `2 !in [2,3]`

**Comment by StefanKarpinski** [comment]
Allowing `!` before infix word operators (`in`, `isa`) seems like it would ok to me. However it's slightly dangerous territory since by analogy if `a !in b` means `!(a in b)` then one would expect that `a !== b` to mean `!(a == b)` but it actually means `!(a === b)`.

**Comment by Moelf** [comment]
maybe disallow `!==` say ambiguity error?

**Comment by yurivish** [comment]
You also can't broadcast `in` in infix form, though I think it would be nice if you were able to and you can already do this with `∈`.

```
julia> in.(1, [[1, 2], [3, 4]])
2-element BitArray{1}:
  true
 false

julia> 1 .in [[1, 2], [3, 4]]
ERROR: syntax: space before "." not allowed in "1 ."

julia> 1 .∈ [[1, 2], [3, 4]]
2-element BitArray{1}:
  true
 false
```

**Comment by StefanKarpinski** [comment]
> maybe disallow `!==` say ambiguity error?

Yeah, to hell with no breaking changes 😂

**Comment by Moelf** [comment]
oh... right `!==` is already there sorry

**Comment by nvhoang3110** [comment]
Please is there any news regarding to the issue ? Some think like` not in` just like in Python would be nice

**Comment by Moelf** [comment]
I feel like the best we can do is special case `notin` as an infix thing, similar to `isa`
